source/elements/oneMKL/source/domains/dft/config_params/storage_formats.rst

```diff
@@ -7,273 +7,227 @@
+encapsulating the real and imaginary parts, separately. The real (resp.
+imaginary) part of any relevant complex element
+:math:`z^{m}_{k_1, k_2,\dots ,k_d}` within the :math:`m`-th
+:math:`d`-dimensional sequence is accessed/stored from/in a data container
+provided at compute time at the index value expressed in eq.
+:eq:`eq_idx_data_layout` (from :ref:`this page<onemkl_dft_config_data_layouts>`)
+of that data container whose elementary data type is ``float`` (resp. ``double``)
+for descriptors operating in single (resp. double)
+:ref:`precision<onemkl_dft_enum_precision>`.
+
+The following snippet illustrates the usage of ``config_value::REAL_REAL``
+storage for the in-place, single-precision (fp32), calculation of :math:`M`
+three-dimensional :math:`n_1 \times n_2 \times n_3` complex transforms, using
+identical (default) strides and distances in forward and backward domains, with
+USM allocations.
 
 .. code-block:: cpp
 
-   datatype * AR; // sizeof(datatype)*n1*n2*n3*m
-   datatype * AI; // sizeof(datatype)*n1*n2*n3*m
-   std::vector<std::int64_t> ios; // length 4 of strides
-   std::int64_t iodist;
-
-   // ...
-
-   // on input: Z(k1,k2,k3,m)
-   //   =   AR[ ios[0] + k1*ios[1] + k2*ios[2] + k3*ios[3] + m*iodist ]
-   //   + i*AI[ ios[0] + k1*ios[1] + k2*ios[2] + k3*ios[3] + m*iodist ]
-   compute_forward(descr, AR, AI); // complex-to-complex in-place FFT
-   // on output: Z{k1,k2,k3,m} 
-   //   =   AR[ ios[0] + k1*ios[1] + k2*ios[2] + k3*ios[3] + m*iodist ]
-   //   + i*AI[ ios[0] + k1*ios[1] + k2*ios[2] + k3*ios[3] + m*iodist ]
-
-
-.. _onemkl_dft_real_storage:
-
-REAL_STORAGE
-++++++++++++
-
-For the :ref:`onemkl_dft_enum_domain` template parameter with value ``REAL``, only the value of ``REAL_REAL`` is supported.
+    namespace dft = oneapi::mkl::dft;
+    dft::descriptor<dft::precision::SINGLE, dft::domain::COMPLEX> desc({n1, n2, n3});
+    std::vector<std::int64_t> strides({0, n2*n3, n3, 1});
+    std::int64_t dist = n1*n2*n3;
+    float *AR = (float *) malloc_device(sizeof(float)*n1*n2*n3*M, queue); // data container for real parts
+    float *AI = (float *) malloc_device(sizeof(float)*n1*n2*n3*M, queue); // data container for imaginary parts
+    desc.set_value(dft::config_param::FWD_STRIDES, strides);
+    desc.set_value(dft::config_param::BWD_STRIDES, strides);
+    desc.set_value(dft::config_param::FWD_DISTANCE, dist);
+    desc.set_value(dft::config_param::BWD_DISTANCE, dist);
+    desc.set_value(dft::config_param::NUMBER_OF_TRANSFORMS, M);
+    desc.set_value(dft::config_param::COMPLEX_STORAGE, dft::config_value::REAL_REAL);
+    desc.commit(queue);
+
+    // initialize forward-domain data such that the real part of z{m;k1,k2,k3}
+    //   = AR[ strides[0] + k1*strides[1] + k2*strides[2] + k3*strides[3] + m*dist ]
+    // and the imaginary part of z{m;k1,k2,k3}
+    //   = AI[ strides[0] + k1*strides[1] + k2*strides[2] + k3*strides[3] + m*dist ]
+    compute_forward(desc, AR, AI); // complex-to-complex in-place DFT
```

Comment:
I'm afraid I'm a little confused by this comment and/or because the ``compute_*`` APIs themselves are actually confusing and possibly identical in those two cases: their specialization, even if explicitly written down, could be the exact same (possible implicit re-interpretation done by the descriptor if not-inplace COMPLEX-COMPLEX transform).
The differences in those two cases is reflected (entirely) by the (complex) descriptor's configuration values set for configuration parameters ``config_param::PLACEMENT`` and ``config_param::COMPLEX_STORAGE``. The latter is explicitly set in this example and I relied on the default value for the former. Maybe it would be clearer to explicitly set it nonetheless? Would you be in favor of that?

Reply:
Ye, those function signatures are a bit of a mess which is why I commented.
The current open-source oneMKL implementation doesn't look at the config_params when you call compute_*, it just assumes you've called the correct function (maybe that is bad), so I'm more thinking about the user accidentally calling the wrong function rather than your intentions being clear.
I think it would be good to explicitly specialize the call.

Reply:
I do not follow/understand what is suggested here, I'm sorry. Using namespace `oneapi::mkl::dft`, the specialized call would be
`compute_forward<descriptor<domain::COMPLEX, precision::SINGLE>,float,float>(...)`
in this example. This would 1. not clarify anything as that _could_ be used for "not-inplace COMPLEX-COMPLEX transform" as well (invoking/triggering an implicit re-interpretation of I/O data) 2. make the example unnecessarily verbose in my opinion.

I agree that it would be more desirable to design the specs in a way that prevents "accidental misuse" of the compute functions but I would argue that this would require an even deeper revision of the specs which falls beyond the scope of this targeted PR, in my opinion (said scope that's already been redefined from "adding fwd/bwd strides" to make it threefold "1. adding fwd/bwd strides, 2. making everything consistent, 3. cleaning up unused or nonsensical options.").

I would need an actual suggestion as to how you would like to see this example updated if that is not what you meant, please.

Reply:
I agree it's not the time to update the `compute_xxxward` functions.

```compute_forward<descriptor<domain::COMPLEX, precision::SINGLE>,float,float>(...)```

would explicitly be a not-inplace COMPLEX-COMPLEX transform since in-place REAL-REAL `compute_forward` only has 2 template parameters.

I'm suggesting
```
compute_forward<decltype(desc), float>(...)
```
I'm a little hazy on template overload resolution, but I believe the in-place REAL-REAL is "more specialized" here since there are fewer template parameters, so would be chosen.

Reply:
Aaaaah, thank you so much: I am guilty of an oversight on the compute functions! I had not realized that. I took your suggestion and applied it in [76ab2a8157de9817783415136328bcdab781cabf](https://github.com/oneapi-src/oneAPI-spec/pull/506/commits/76ab2a8157de9817783415136328bcdab781cabf)